pulp_rpm/app/tasks/publishing.py

```diff
@@ -117,18 +117,46 @@ def publish_artifacts(self, content, prefix=""):
             content__pulp_type=Package.get_pulp_type()
         )
 
+        paths = set()
+        duplicated_paths = []
         for content_artifact in contentartifact_qs.values("pk", "relative_path").iterator():
             relative_path = content_artifact["relative_path"]
             relative_path = os.path.join(
                 prefix, PACKAGES_DIRECTORY, relative_path.lower()[0], relative_path
             )
+            #
+            # Some Suboptimal Repos have the 'same' artifact living in multiple places.
+            # Specifically, the same NEVRA, in more than once place, **with different checksums**
+            # since if all that was different was location_href there's be only one ContentArtifact
+            # in the first place). In this case, we pick one - which is essentially what the rest
```

Comment:
So there are multiple `ContentArtifact` for one `Content`?  I'm not sure I understand this part.

The `Package` content includes `pkgId` which is a checksum so I would think that the content part, at least, wouldn't ever be shared.

Or are we really just using `ContentArtifact` as a proxy for `Content`, and the 1-to-1 invariant still holds, but it's 2 separate packages.  Now that I'm typing this it sounds more like the latter :)

Reply:
At this point, we are dealing with Packages, explicitly, and have found "all the ContentArtifacts that correspond to the Content-set that was passed in, that are type=Package". In this edge-case, we have (at least) two ContentArtifacts, which point at different Artifacts (because their checksums are different), but have the **same** content-artifact.relative_path (becxause the package-nevra/filename is the same). To publish, we have to pick one as the "winner" that will show up at Packages/p/pkg-1.2.3-4.x86_64.rpm

Does that help any?

Reply:
I think so.  What I was trying to verify is that we weren't somehow getting multiple ContentArtifacts *pointing at one Package*.  And it sounds like that particular invariant wasn't being violated, we're talking about multiple packages here, each with exactly one ContentArtifact.